index.js

```diff
@@ -1,77 +1,184 @@
-var os = require('os')
-var fs = require('fs')
-var path = require('path')
-var Cache = require('ttl')
+const fs = require('fs')
+const path = require('path')
+const Xache = require('xache')
+const safetyCatch = require('safety-catch')
 
-var isLinux = os.platform() === 'linux' // native recursive watching not supported here
-var watchDirectory = isLinux ? watchFallback : watchRecursive
+// Native recursive watching is not supported on Linux based platforms
+const isLinux = process.platform === 'linux' || process.platform === 'android'
+const watchDirectory = isLinux ? watchFallback : watchRecursive
 
-module.exports = watch
+module.exports = class RecursiveWatch {
+  constructor (filename, onchange) {
+    this.filename = filename
+    this.onchange = onchange
 
-function watch (name, onchange) {
-  var clear = null
-  var stopped = false
+    this.unwatch = null
 
-  fs.lstat(name, function (_, st) {
-    if (!st || stopped) {
-      stopped = true
-      return
+    this.opened = false
+    this.closed = false
+
+    this._opening = this._ready()
```

Comment:
Ready Resource? Then you don't need to worry about all this plumbing

Reply:
Very true! ReadyResource has `closing` and `closed` so it should be good